rfcs/0012-pipeline-executor.md

```diff
@@ -0,0 +1,367 @@
+<!--- Licensed to the Apache Software Foundation (ASF) under one -->
+<!--- or more contributor license agreements.  See the NOTICE file -->
+<!--- distributed with this work for additional information -->
+<!--- regarding copyright ownership.  The ASF licenses this file -->
+<!--- to you under the Apache License, Version 2.0 (the -->
+<!--- "License"); you may not use this file except in compliance -->
+<!--- with the License.  You may obtain a copy of the License at -->
+
+<!---   http://www.apache.org/licenses/LICENSE-2.0 -->
+
+<!--- Unless required by applicable law or agreed to in writing, -->
+<!--- software distributed under the License is distributed on an -->
+<!--- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY -->
+<!--- KIND, either express or implied.  See the License for the -->
+<!--- specific language governing permissions and limitations -->
+<!--- under the License. -->
+- Feature Name: (fill me in with a unique identifier, `my_awesome_feature`)
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [apache/tvm-rfcs#0014](https://github.com/apache/tvm-rfcs/pull/0014)
+- GitHub Issue: [apache/tvm#8596](https://github.com/apache/tvm/issues/8596)
+
+## 1. Summary
+
+
+This proposal introduces Pipeline Executor: A runtime executor that by scheduling
+splitted subgraph of relay graph in pipeline to implement task level parallism to
+reduce compute latency.
```

Comment:
about #2, compute latency, in blocking/serialize mode, let's say we get 2 data at same time and need to do compute, for every single data the processing time is 4ms, then first data latency is 4ms, second data latency is 4ms + 4ms = 8ms, in 4 stage pipe line mode,  every stage spend 1ms, then first data latency still is 4ms but second data processing time get reduced into 1ms + 4ms = 5ms, hence we can say latency get reduced.

about #1, yes this RFC only include executor, i will add the expected input part

Reply:
On point 2, with latency in ML we typically mean the time it takes to process one unit of data, which for image classification is one image if the batch size is 1. In the example given, I would argue that the batch size is still 1 instead of 2 as the data inputs are just being queued and executed separately and not being processed together. Therefore, the latency is not reduced but the throughput is improved.

Reply:
about "not being processed together", actually that is not true,  for image processing in pipeline case, when image 1 released stage 1 and doing compute in stage 2 that time, image 2 will do processing in stage 1, in such scenario,  image 1 and image 2  being processed together. 

I agree one unit of data latency is the typically latency in ML, but from my point of view, in deployment case multiple data  latency are more meaningful for product performance evaluation.

Reply:
I actually don't think it's a typical case only in ML. This concept is also widely used in HPC. The "latency" is based on the time of executing one task; while the "throughput" is the number of finished task in a time unit. Pipelining definitely cannot improve latency (of executing one task) but throughput, because you can finish more tasks in the same time period.

General speaking, I think this is just the terminology issue, but it would be good to use consistent terminology for efficient communications. The confusion of "not being processed together" falls into the similar reason. Jorn and probably most people may feel confuse about why two stages of the same task can be processed together, but you actually meant stage 2 of the _first task_ and the stage 1 of the _second task_ can be processed in parallel. This is the point you should make it clear in the RFC.

Reply:
With `not being processed together` I meant that both data inputs are not in the same stage at the same time, which is what is meant with the batch size. You could also execute in a pipelined manner with a batch size > 1, which would mean that you process multiple data inputs at every stage at the same time. So executing with a certain batch size is orthogonal to executing in a pipelined manner and latency refers to the time it takes to process a certain batch of data (typically size 1), irrespective of how you process it (pipelined or otherwise). It's indeed a terminology issue and I think it's best to use the common meaning to avoid confusion, so I think using `increase throughput` instead of `reduce latency` would be better here.

Reply:
@jtuyls, @comaniac, agree, would change into the term "throughput" to avoid confusion.